src/encoder/stateless.rs

```diff
@@ -255,3 +253,167 @@ where
 
     Ok(())
 }
+
+/// Helper aliases for codec and backend specific types
+type Picture<C, B> = <B as StatelessVideoEncoderBackend<C>>::Picture;
+
+type Reference<C, B> = <C as StatelessCodecSpecific<B>>::Reference;
+
+type Request<C, B> = <C as StatelessCodecSpecific<B>>::Request;
+
+type CodedPromise<C, B> = <C as StatelessCodecSpecific<B>>::CodedPromise;
+
+type ReferencePromise<C, B> =
+    <C as StatelessCodecSpecific<B>>::ReferencePromise;
+
+type BoxPredictor<C, B> = Box<dyn Predictor<Picture<C, B>, Reference<C, B>, Request<C, B>>>;
+
+pub struct StatelessEncoder<Codec, Handle, Backend>
+where
+    Backend: StatelessVideoEncoderBackend<Codec>,
+    Codec: StatelessCodec + StatelessCodecSpecific<Backend>,
+{
+    /// Pending frame output promise queue
+    output_queue: OutputQueue<CodedPromise<Codec, Backend>>,
+
+    /// Pending reconstructed pictures promise queue
+    recon_queue: OutputQueue<ReferencePromise<Codec, Backend>>,
+
+    /// [`Predictor`] instance responsible for the encoder decision making
+    predictor: BoxPredictor<Codec, Backend>,
+
+    // predictor: Box<dyn Predictor<B::Picture, B::Reference>>,
+    coded_queue: VecDeque<CodedBitstreamBuffer>,
+
+    /// Number of the currently held frames by the predictor
+    predictor_frame_count: usize,
+
+    /// [`StatelessVP9EncoderBackend`] instance to delegate [`BackendRequest`] to
+    backend: Backend,
+
+    _phantom: std::marker::PhantomData<Handle>,
+}
+
+/// A bridge trait between [`StatelessEncoder`] and codec specific backend trait (eg.
+/// [`h264::StatelessH264EncoderBackend`] or [`vp9::StatelessVP9EncoderBackend`]).
+/// Accepts [`Request`] and is responsible for adding resutling [`BackendPromise`] to
+/// [`StatelessEncoder`] internal queues and  decrementing the internal predictor frame counter if
+/// the backend moved the frame outside predictor ownership.
+pub trait StatelessEncoderExecute<Codec, Handle, Backend>
```

Comment:
Since this has the same constraints has `StatelessEncoder`, and `StatelessEncoder` depends on this trait, can't we just make `execute` a method of `StatelessEncoder`?

Reply:
But in such case I'd have to add constraints for `Handle` and `Backend` as well